torchao/quantization/subgraph_utils/extract_subgraphs.py

```diff
@@ -0,0 +1,632 @@
+import traceback
+import sys
+
+import torch
+import torch.nn.functional as F
+
+from torch.utils._pytree import tree_map
+
+graph_tabular_log = torch._logging.getArtifactLogger(__name__, "graph")
+
+# TODO(future): might be nice to have input shapes here, but needs a refactor since
+# they are easiest to get from the subgraph extraction function, but currently summary
+# is generated from the subgraph debug function.
+summary_headers = ['extraction_idx', 'orig_node_name', 'subgraph_idx', 'lin1_shape', 'lin2_shape', 'subgraph_summary']
+
+
+# A model can have multiple regions torch.compile'd separately. Because we currently
+# depend on torch._inductor.config.pre_grad_custom_pass, that means we will run
+# the extraction logic once per torch.compile'd region. The variable below tracks
+# how many times we have called the top level subgraph extractor, so we can
+# save results from each run without overwriting data.
+DEBUG_LINEARS_CALL_COUNTER = 0
+
+
+def maybe_short_name(torch_fn):
+    """
+    Tries to format things like
+
+      '<built-in method cat of type object at 0x7f3f7a202c60>'
+
+    as
+
+      'torch.cat'
+    """
+    if hasattr(torch_fn, '__name__'):
+        # torch.cat -> cat
+        if hasattr(torch, torch_fn.__name__):
+            if getattr(torch, torch_fn.__name__) == torch_fn:
+                return torch_fn.__name__
+
+        # F.layer_norm -> layer_norm
+        if hasattr(F, torch_fn.__name__):
+            if getattr(F, torch_fn.__name__) == torch_fn:
+                return torch_fn.__name__
+
+        # builtin function mul
+        # note: there is definitely a more generic way to do this
+        if torch_fn.__name__ == 'mul':
+            return 'mul'
+        if torch_fn.__name__ == 'add':
+            return 'add'
+
+        # activation modules
+        # note: there is definitely a more generic way to do this
+        if 'torch.nn.modules.activation' in str(torch_fn):
+            return torch_fn.__name__
+
+    return torch_fn
+
+def get_meta_val(n: torch.fx.Node):
+    # from https://fburl.com/code/hcwdl994
+    meta_val = n.meta.get('val', n.meta.get('tensor_meta', n.meta.get('example_value', None)))
+    return meta_val
+
+def get_stack_summary(n: torch.fx.Node):
+    # from https://fburl.com/code/yify7y7f
+    if n.stack_trace:
+        parsed_stack_trace = torch.fx.graph._parse_stack_trace(n.stack_trace)
+        summary = parsed_stack_trace.get_summary_str()
+        return summary
+    return None
+
+def is_first_node_of_dual_linear(gm: torch.fx.GraphModule, n: torch.fx.Node):
+    first_user = list(n.users.items())[0][0]
+    if first_user.op == 'call_module':
+        first_user_mod = getattr(gm, first_user.target)
+        if type(first_user_mod) == torch.nn.Linear:
+            return True
+    elif first_user.op == 'call_function':
+        if first_user.target == torch._C._nn.linear:
+            return True
+    return False
+
+def debug_single_linear(
```

Comment:
depending on what you're using this file for, this function also might be deleteable

Reply:
yeah will refine more when it's closer to land, right now just experimenting to see if this approach helps improve things over the original approach on models we care about